### PR TITLE
feat(web-simulator): Add a web based simulator for testing the app

### PR DESCRIPTION
**Pull request by pvinis**
https://user-images.githubusercontent.com/100233/224519180-d9bae832-d9d4-4980-b251-a4ad19a881c8.mov

### CODE REVIEW

File: simulator-web/src/utils/storage.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+const stringToUint8Array = (str: string) => {
+  const buf = new ArrayBuffer(str.length)
+  const bufView = new Uint8Array(buf)
+  for (let i = 0, strLen = str.length; i < strLen; i++) {
+    bufView[i] = str.charCodeAt(i)
+  }
+  return bufView
+}
+
+export const useLocalStorageUint8Array = (key: string, initialValue: Uint8Array | undefined) =>
```
Comment (pvinis) [inline simulator-web/src/utils/storage.ts:16]: storing the pub and priv keys in localstorage

File: simulator-web/src/utils/sodium.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import _sodium from "libsodium-wrappers"
+import { useEffect, useRef, useState } from "react"
+
+const libsodium = async () => {
+  await _sodium.ready
+  return _sodium
+}
+
+export const useSodium = (): undefined | typeof _sodium => {
```
Comment (pvinis) [inline simulator-web/src/utils/sodium.ts:9]: nice sodium hook

File: simulator-web/src/Canvas.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  const canvasRef = useRef<HTMLCanvasElement>(null)
+
+  useEffect(() => {
+    const canvas = canvasRef.current!
+    const ctx = canvas.getContext("2d")!
+
+    const imageData = ctx.getImageData(0, 0, 640, 384)
+    imageData.data.set(pixels)
+
+    ctx.putImageData(imageData, 0, 0)
```
Comment (pvinis) [inline simulator-web/src/Canvas.tsx:18]: thanks again for this!

File: simulator-web/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import QRCode from "react-qr-code"
+import { map, pipe } from "remeda"
+import { Canvas } from "./Canvas"
+import { decodeBundle } from "./utils/decodeBundle"
+import { expand2BPPValues } from "./utils/expand2BPP"
+import { useSodium } from "./utils/sodium"
+import { useLocalStorageUint8Array } from "./utils/storage"
+import { RLEDecoder } from "./utils/RLEDecoder"
+
+export const App = () => {
```
Comment (pvinis) [inline simulator-web/src/App.tsx:11]: a nice first version react simulator

**Comment by jbmorley** [comment]
@pvinis This is looking great! I think we should probably merge it as is and then iterate. How does that sound?

**Review by jbmorley — APPROVED** [review]

**Comment by jbmorley** [comment]
@pvinis I've updated this branch to make sure it can still be merged. There are a couple of things I think we should do to this code pretty quickly but they're not urgent for this PR--feel free to do them here if you're excited 🎉.

- [ ] Update to the new v3 URLs. I'd forgotten that I was half way through trying to clean up the APIs which didn't correctly namespace updates and device registration. The update URL should be `/api/v3/status/<device-identifier>`
- [ ] Switch to generating and saving per-browser actual UUIDs (https://en.wikipedia.org/wiki/Universally_unique_identifier); the service now accepts canonical format UUIDs as well last the older 8-character format

File: simulator-web/src/App.tsx (hunk 4 above)
Comment (jbmorley) [reply simulator-web/src/App.tsx:11]: 🥳

File: simulator-web/src/Canvas.tsx (hunk 3 above)
Comment (jbmorley) [reply simulator-web/src/Canvas.tsx:18]: Glad it worked so smoothly.

File: simulator-web/src/utils/storage.ts (hunk 1 above)
Comment (jbmorley) [reply simulator-web/src/utils/storage.ts:16]: Nice. I think this is where we'll want the per-instance UUID too.

**Pull request merged**